torch/_dynamo/variables/misc.py

```diff
@@ -116,6 +116,68 @@ def exit(self, tx, *args):
         self._call_func(tx, self.initial_values)
         return variables.ConstantVariable(None, **VariableTracker.propagate(self))
 
+    def _call_func(self, tx, initial_values):
+        raise NotImplementedError("_call_func called on base")
+
+    def _func_name(self):
+        raise NotImplementedError("_func_name called on base")
+
+    def call_function(
+        self, tx, args: "List[VariableTracker]", kwargs: "Dict[str, VariableTracker]"
+    ) -> "VariableTracker":
+        assert len(args) == 1
+        assert isinstance(args[0], UserMethodVariable) or isinstance(
+            args[0], UserFunctionVariable
+        )
+
+        if isinstance(args[0], UserMethodVariable):
+            return WrappedUserMethodVariable(args[0], self)
+
+        if isinstance(args[0], UserFunctionVariable):
+            return WrappedUserFunctionVariable(args[0], self)
+
+
+class GradModeVariable(ContextWrappingVariable):
+    """represents torch.{no_grad,enable_grad,set_grad_mode}()"""
+
+    _guards_singleton = {Guard("", GuardSource.GLOBAL, GuardBuilder.GRAD_MODE)}
+
+    @staticmethod
+    def create(tx, target_value, **kwargs):
+        var = GradModeVariable(
+            target_values=[target_value],
+            initial_values=[torch.is_grad_enabled()],
+            **kwargs,
+        )
+        var._call_func(tx, [target_value])
+        return var
+
+    def __init__(self, target_values, initial_values=None, **kwargs):
+        super(GradModeVariable, self).__init__(
+            target_values=target_values, initial_values=initial_values, **kwargs
+        )
+        self.guards = self.guards | self._guards_singleton
+
+    def enter(self, tx):
+        return variables.ConstantVariable(None, **VariableTracker.propagate(self))
+
+    def _call_func(self, tx, values):
+        assert len(values) == 1
+        value = values[0]
+        tx.output.graph.create_node(
+            "call_function", torch._C._set_grad_enabled, (value,), {}
+        ),
+        torch._C._set_grad_enabled(value)
+
+    def _func_name(self):
+        return "_C._set_grad_enabled"
+
+    def fn_name(self):
+        if self.target_values[0]:
```

Comment:
For reviewer: this line is the only real change. After the refactor at  https://github.com/pytorch/torchdynamo/pull/1289, ```target_values``` becomes a list, so we should get the grad state from ```[0]```. Otherwise, this line always return true.
For the other part, I just move function ```reconstruct``` from the parent class ```ContextWrappingVariable``` to ```GradModeVariable```, as this function's logic is used to handle grad mode specially ATM. Github seems doesn't show diff very well.